File: workspace/agent/session.go
package main

import (
	"fmt"
	"path/filepath"
	"strings"
	"time"

	"github.com/k-k1/agent-fleet/workspace/agent/internal/agents"
	"github.com/k-k1/agent-fleet/workspace/agent/internal/session"
	"github.com/k-k1/agent-fleet/workspace/agent/internal/status"
)

// セッションのワイヤ変換とタイトル/ラベル導出。モデル・メタ永続化・UUID は
// internal/session（docs/23 残① Wave A）/ tmux= session_tmux.go /
// HTTPハンドラ= session_handlers.go / claude の CLI 起動コマンドは
// internal/agents/claude の program.go（docs/23 残① Wave F）

// wireSession builds the API representation from a meta and liveness.
func wireSession(m session.Meta, alive bool) session.Session {
	started := ""
	if m.CreatedAt != "" {
		if t, err := time.Parse(time.RFC3339, m.CreatedAt); err == nil {
			started = t.Local().Format("01/02 15:04")
		}
	}
	// The live-dependent fields (state / remote URL / context / resumable / bg-busy)
	// diverge by kind — the agent computes them (see WireLive per implementation).
	li := agentOf(m.Kind).WireLive(m, alive)
	s := session.Session{
		Name: m.Name, Tmux: session.TmuxName(m.Name), Dir: m.Dir, Subdir: m.Subdir, Kind: m.Kind, Driver: m.Driver,
		Repo: m.Repo, WorkingCopyID: workingCopyID(m.Dir), Title: m.Title, Display: session.Display(m), Color: m.Color, Label: m.Label,
		Started: started, CreatedAt: m.CreatedAt, Branch: m.Branch,
		RemoteUrl: li.RemoteURL, State: li.State, Alive: alive, Resumable: li.Resumable,
		BackgroundBusy: li.BackgroundBusy, Context: li.Context, Locked: m.Locked, Archived: m.Archived,
	}
	// 上限で切れたターンの後始末が済んだ claude（メニューは自動解除済み／モデル別上限は
	// そもそもメニューを出さない）はペインが待機プロンプトに戻るので、ここまでの状態は
	// idle＝入力待ちになる。リセットを待っているだけなのに正常終了と見分けが付かない
	// ので、開いているエピソードがある間は 制限解除待ち として名乗る（docs/47 §4-9）。
	// WireLive ではなくここで見るのはエピソードの持ち主が package main だからで、
	// driveState（チャット／ミラーのチップ）にも同じ読み替えが要る。
	if alive && s.State == "idle" && normalizeKind(m.Kind) == session.KindClaude {
		if at, waiting := rateLimitWaiting(m, time.Now()); waiting {
			s.State = agents.StateLimited
			s.RateLimitResumeAt = at
		}
	}
	// For a stopped session, surface WHY it ended (crash / OOM) if the pane recorder
	// captured a cause. A clean quit (exited) or a deliberate stop (empty reason) leaves
	// the fields unset so the row shows the plain 停止中 chip.
	if !alive {
		if e, ok := status.ReadExit(m.Name); ok && e.Reason != "" && e.Reason != "exited" && e.Reason != "stopped" {
			s.ExitReason = e.Reason
			s.ExitCode = e.Code
			s.ExitSignal = e.Signal
		}
	}
	return s
}

// remoteSessionURL（claude.ai Remote Control URL の導出）は internal/agents/claude
// の claude.RemoteSessionURL へ移設（docs/23 残① Wave F）。

// dirInfo is a working copy's current branch + worktree flag, cached per dir.
type dirInfo struct {
	branch   string
	worktree bool
}

// annotateSessions enriches each session from its working copy: Worktree (is it a
// linked worktree) and BranchDrift/CurrentBranch (was it switched off its start
// branch). info resolves a dir once and is cached, so N sessions sharing a working
// copy cost a single git call. Drift needs a recorded start branch; worktree does not.
// Split from the git/tmux plumbing so it is unit-testable.
func annotateSessions(sessions []session.Session, info func(string) dirInfo) {
	cache := map[string]dirInfo{}
	for i := range sessions {
		s := &sessions[i]
		if s.Dir == "" {
			continue
		}
		v, ok := cache[s.Dir]
		if !ok {
			v = info(s.Dir)
			cache[s.Dir] = v
		}
		s.Worktree = v.worktree
		if s.Branch != "" && v.branch != "" && v.branch != s.Branch {
			s.BranchDrift = true
			s.CurrentBranch = v.branch
		}
	}
}

// sessionLabelFor builds the claude --name for a session. When the user supplied a
// title it's "[AF] {title}"; otherwise it falls back to the auto default
// "[AF] {repo} @MMDD-HHMM" where {repo} is the working dir's basename and the time is
// the workspace's local time (the entrypoint exports TZ from the per-user timezone
// setting, default JST). The "[AF] " tag identifies Agent-Fleet-launched sessions in
// the claude.ai Remote Control picker. Computed at create/recreate and stored in the
// meta so relaunch keeps the same name.
func sessionLabelFor(dir, title string) string {
	if title != "" {
		return "[AF] " + title
	}
	return fmt.Sprintf("[AF] %s @%s", filepath.Base(dir), time.Now().Format("0102-1504"))
}

// cleanTitle trims and validates a user-supplied display title. It rejects control
// characters (which would corrupt the tmux title / claude --name) and caps the length.
// An empty title is valid (the session uses the auto default label). Returns ok=false
// only for an over-long or control-laden title.
func cleanTitle(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if len([]rune(s)) > 80 {
		return "", false
	}
	for _, r := range s {
		if r < 0x20 || r == 0x7f {
			return "", false
		}
	}
	return s, true
}

// forkTitle derives a fork's display title from its source: the source's own title
// when set, else its stripped label (the auto "{repo} @time"), suffixed " (fork)".
func forkTitle(src session.Meta) string {
	base := src.Title
	if base == "" {
		base = strings.TrimPrefix(src.Label, "[AF] ")
	}
	return strings.TrimSpace(base + " (fork)")
}

// shellQuote は internal/session の session.ShellQuote へ移設（docs/23 残① Wave D）。

File: console/src/types/session.ts
// Session domain types. A "session" is one tmux slot the Agent runs (a coding
// agent or a plain shell / an SSM login). `kind` is the discriminator the whole
// Console keys per-agent behavior on — presentation, availability, launch params,
// and capabilities all derive from it via the agent registry (src/agents/registry).
//
// NOTE: this session `kind` is a DIFFERENT axis from a *pane's* kind (which VIEW
// renders — terminal/file/scm/doc/diff, see layout/types PaneContent). Keep distinct.

export type SessionKind = "claude" | "codex" | "cursor" | "copilot" | "kiro" | "opencode" | "agy" | "shell" | "ssm";

// The canonical session kinds in display order (New Session buttons, etc.).
export const SESSION_KINDS: SessionKind[] = ["claude", "codex", "cursor", "copilot", "kiro", "agy", "opencode", "shell", "ssm"];

// Live run state, reported by per-agent hooks/plugins. "" (empty) = idle. claude
// emits question/plan/permission; codex/opencode emit working/idle; agy emits
// question/permission only (conversation-DB probe — no working/idle hooks);
// shell/ssm emit nothing (their liveness is shown from `alive`).
export type SessionState = "" | "working" | "idle" | "question" | "plan" | "permission";

// A session as returned by GET /api/sessions and used across the left pane, the
// terminal header, and the chat mirror. Optional fields may be absent per kind.
export interface Session {
  name: string; // auto-allocated unique slug ("s" + 6 base32 chars, e.g. "sukbq4s") — the session's immutable identity
  kind: SessionKind;
  // Control route (docs/log/27): "managed" = shared runtime + structured RPC (AF is the only
  // writer, no tmux pane). absent/"tui" = the classic TUI inside tmux. Whether a pane exists
  // is decided by this axis, not by kind — always branch through isManagedSession().
  driver?: "tui" | "managed" | string;
  title?: string; // user-supplied display title (optional, any kind); "" = auto
  color?: string; // terminal background hue (hex); SSM sessions carry their host color
  label?: string; // claude --name (with an "[AF] " tag); absent for shell
  repo?: string | null; // working-copy folder the (agent) session runs in
  workingCopyId?: string;
  path?: string; // absolute working dir
  dir?: string; // working dir shown in the row tooltip
  // Folder BENEATH dir the agent actually runs in, slash-relative ("" / absent = dir
  // itself). Sessions stay grouped by `dir` (the working copy) — this is only the
  // extra "where inside it" detail.
  subdir?: string;
  remoteUrl?: string; // clone URL (agent sessions with a repo)
  state?: SessionState | string; // live hook/plugin state ("" = idle)
  alive?: boolean; // tmux session is running
  resumable?: boolean; // a stopped session whose dir still exists (false = archive only)
  backgroundBusy?: boolean; // idle by hook but a run_in_background task is still running
  backgroundBusyReason?: string;
  // The reserved auto-resume instant (RFC3339), present only while state === "limited"
  // (waiting for a usage limit to reset). Empty = no resume is scheduled (auto-resume off,
  // nothing to derive a time from, or a per-model limit). Display-only, so the chip can say
  // when the session moves again (docs/log/47 §4-9).
  rateLimitResumeAt?: string;
  createdAt?: string; // ISO timestamp
  model?: string; // claude model
  context?: SessionContextUsage; // claude context-window usage (the Agent's session.ContextUsage)
  branch?: string; // git branch the working copy was on when the session started
  currentBranch?: string; // working copy's branch now, set only when it differs from `branch`
  branchDrift?: boolean; // true = the working tree was switched off `branch` under the session
  worktree?: boolean; // session runs in a linked git worktree (offers branch rename)
  // Why a STOPPED session's agent process ended, when it ended abnormally: "oom"
  // (memory-killed), "killed" (SIGKILL, non-OOM) or "crashed" (fault / non-zero exit).
  // Absent for live sessions, clean quits, and deliberate stops. exitCode is the raw
  // pane wait status (137 = OOM SIGKILL); exitSignal the derived signal number.
  exitReason?: "oom" | "killed" | "crashed" | string;
  exitCode?: number;
  exitSignal?: number;
  // Which kind of interaction was still awaiting an answer when the session was stopped
  // (docs/log/75). Only set on stopped rows, where it turns the list badge into "stopped,
  // question pending". While the session is alive, `state` says the same thing.
  carried?: "question" | "plan" | "permission" | string;
  // Deletion lock (docs/log/45): while true, the Agent answers 403 to anything that deletes
  // (delete = forget the metadata, purge, the 7-day auto-prune of stopped sessions, and
  // removal as a side effect of deleting the working copy). Stop and archive are reversible
  // and still go through. The row's lock badge and the disabled delete items read this.
  locked?: boolean;
  // Keep-awake pin (docs/log/75): exempt from the idle auto-stop until this instant.
  // Past or empty = no pin is held.
  keepAwakeUntil?: string;
}

// A session's current context fill — the wire shape of the Agent's
// session.ContextUsage (internal/session/session.go). Field names deliberately match
// the ContextBar props so the terminal/chat heads can spread it straight in.
export interface SessionContextUsage {
  read: number; // cache_read_input_tokens (cache reuse)
  create: number; // cache_creation_input_tokens (new cache)
  fresh: number; // input_tokens (uncached)
  model?: string;
}

// isManagedSession: a managed (paneless) session has no tmux pane — the chat
// mirror is its primary UI, no terminal view exists, and its inputs go through
// the semantic /turn and /respond ops instead of TUI key driving (docs/log/27 §10).
export const isManagedSession = (s?: { driver?: string } | null): boolean =>
  s?.driver === "managed";

// Provider connection status for one agent, from GET /api/connections.
export interface ProviderConn {
  connected?: boolean;
  envs?: string[]; // opencode: configured provider API-key env names (auth.go)
  // agy: host capability (docs/log/32 Track B — the RDRAND guard). false = this host
  // cannot run agy ("no_rdrand" / "not_installed"); absent = supported.
  supported?: boolean;
  reason?: string;
  // Chat integrations (discord / slack): the display form of the notification master switch
  // (the inverse of notifyOff). OFF only when false is explicit — unset (an older
  // connection) counts as ON.
  notify?: boolean;
  // claude (docs/log/47 §4-8): the OAuth credential's expiry. `claude auth status` never
  // reports an expiry, so the Agent reads refreshTokenExpiresAt out of the credential itself
  // and puts it here. expired = both the refresh and the access token are past (no turn can
  // start any more); days_left = an advance warning, present only within 3 days of expiry.
  // Their absence means there is nothing to judge from (API-key operation, not connected, a
  // changed format) — it does NOT mean expired.
  expires_at?: string;
  expired?: boolean;
  days_left?: number;
  // opencode: the selected billing route (docs/log/54). "free" is a tier that launches with
  // no authentication at all, so the launch gate reads this and allows opencode even when
  // not connected. "off" is the opposite: an explicit disable that closes the launch gate
  // even when a key or OAuth exists (connected becomes false).
  usage?: "off" | "free" | "go" | "zen";
  // opencode: an opencode account connection, which coexists with API keys (envs) (docs/log/54).
  oauth?: boolean; // connected through the Console account
  oauth_label?: string; // organisation name of the connection (the label opencode returns)
  oauth_known?: boolean; // false = unverified because the serve daemon is not up (not necessarily disconnected)
  oauth_disabled?: boolean; // managed opencode is disabled, so no sign-in route can be offered
  // opencode: a link to the usage page (opencode.ai/workspace/{id}/go). There is no API for
  // the numbers, so this holds only the ID, the URL, and whatever quota information could be
  // observed when a limit was hit.
  workspace_id?: string;
  workspace_id_source?: "manual" | "learned";
  workspace_url?: string;
  last_limit?: { name?: string; reset_at?: string };
}

// The full connections bag. Known agents are named; git providers (github /
// bitbucket) and any future entries ride the index signature.
export interface ConnectionsStatus {
  claude?: ProviderConn;
  codex?: ProviderConn;
  // cursor (docs/log/40): its own login flow. connected = ~/.config/cursor/auth.json exists;
  // supported=false = an older image without the CLI baked in.
  cursor?: ProviderConn;
  opencode?: ProviderConn;
  agy?: ProviderConn;
  // copilot rides on the GitHub integration (docs/log/36): connected = a gh token exists;
  // supported=false = an older image without the CLI baked in.
  copilot?: ProviderConn;
  // kiro (docs/log/43): device-flow login. connected = credentials exist (whoami exit 0);
  // supported=false = the CLI is not installed (before on-demand install; ~855MB goes to the
  // per-user home).
  kiro?: ProviderConn;
  [provider: string]: ProviderConn | undefined;
}

// A registered SSM host bookmark, from GET /api/ssm/hosts.
export interface SsmHost {
  id: string;
  alias: string;
  profileId: string;
  region: string;
  instanceId: string;
  documentName: string;
}
